src/main/java/net/fabricmc/loader/impl/discovery/ResultAnalyzer.java

```diff
@@ -0,0 +1,287 @@
+				getDependencyVersionRequirements(dependency), dependency.getModId());
+	}
+
+	private static void appendUnsatisfiedDependencyError(ModDependency dependency, List<ModCandidate> matches, PrintWriter pw) {
+		pw.printf("but a non-matching version is present: %s!\n\t - You must install %s of %s.",
+				getCandidateFriendlyVersions(matches), getDependencyVersionRequirements(dependency), getCandidateName(matches.get(0)));
+	}
+
+	private static void appendConflictError(ModCandidate candidate, List<ModCandidate> matches, PrintWriter pw) {
+		final String depCandidateVer = getCandidateFriendlyVersions(matches);
+
+		pw.printf("but a matching version is present: %s!\n"
+				+ "\t - While this won't prevent you from starting the game,  the developer(s) of %s have found that "
+				+ "version %s of %s conflicts with their mod.\n"
+				+ "\t - It is heavily recommended to remove one of the mods.",
+				depCandidateVer, getCandidateName(candidate), depCandidateVer, getCandidateName(matches.get(0)));
+	}
+
+	private static void appendBreakingError(ModCandidate candidate, List<ModCandidate> matches, PrintWriter pw) {
+		final String depCandidateVer = getCandidateFriendlyVersions(matches);
+
+		pw.printf("but a matching version is present: %s!\n"
+				+ "\t - The developer(s) of %s have found that version %s of %s critically conflicts with their mod.\n"
+				+ "\t - You must remove one of the mods.",
+				depCandidateVer, getCandidateName(candidate), depCandidateVer, getCandidateName(matches.get(0)));
+	}
+
+	private static void appendJiJInfo(List<ModCandidate> mods, PrintWriter pw) {
+		for (ModCandidate mod : mods) {
+			if (mod.getMinNestLevel() < 1) {
+				pw.printf("\n\t - Mod %s v%s is being loaded from the mods directory.",
+						getCandidateName(mod), getCandidateFriendlyVersion(mod));
+				return;
+			}
+
+			List<ModCandidate> path = new ArrayList<>();
+			ModCandidate cur = mod;
+
+			do {
+				for (ModCandidate parent : cur.getParentMods()) {
+					if (parent.getMinNestLevel() < cur.getMinNestLevel()) {
+						path.add(parent);
+						cur = parent;
+						break;
+					}
+				}
+			} while (cur.getMinNestLevel() > 0);
+
+			StringBuilder pathSb = new StringBuilder();
+
+			for (int i = path.size() - 1; i >= 0; i--) {
+				ModCandidate m = path.get(i);
+
+				if (pathSb.length() > 0) pathSb.append(" -> ");
+				pathSb.append(String.format("%s v%s", getCandidateName(m), getCandidateFriendlyVersion(m)));
+			}
+
+			// now we have the proper data, yay
+			pw.printf("\n\t - Mod %s v%s is being provided through e.g. %s.",
+					getCandidateName(mod), getCandidateFriendlyVersion(mod), pathSb);
+		}
+	}
+
+	private static String getCandidateName(ModCandidate candidate) {
+		return String.format("'%s' (%s)", candidate.getMetadata().getName(), candidate.getId());
+	}
+
+	private static String getCandidateFriendlyVersion(ModCandidate candidate) {
+		return candidate.getVersion().getFriendlyString();
```

Comment:
Will (the use of) this result in errors like `The developer(s) of x_mod have found that version * of y_mod critically conflicts with their mod.`? If so, where the version is wildcarded could it be made clearer along the lines of `The developer(s) of x_mod have found that y_mod critically conflicts with their mod.`, so that it is more apparent there is no version which could ever work?

Reply:
I don't think so, ModCandidate is always a concrete mod, not a dependency reference.